avro4s-core/src/main/scala/com/sksamuel/avro4s/AvroValue.scala

```diff
@@ -0,0 +1,63 @@
+package com.sksamuel.avro4s
+
+import java.nio.ByteBuffer
+
+import org.apache.avro.generic.{GenericEnumSymbol, GenericFixed, GenericRecord}
+import org.apache.avro.util.Utf8
+
+sealed trait AvroValue
+
+object AvroValue {
+
+  import scala.collection.JavaConverters._
+
+  case class AvroString(s: String) extends AvroValue
+  case class AvroByteArray(bytes: Array[Byte]) extends AvroValue
+  case class AvroGenericFixed(fixed: GenericFixed) extends AvroValue
+  case class AvroBoolean(boolean: Boolean) extends AvroValue
+  case class AvroShort(short: Short) extends AvroValue
+  case class AvroByte(byte: Byte) extends AvroValue
+  case class AvroDouble(double: Double) extends AvroValue
+  case class AvroFloat(float: Float) extends AvroValue
+  case class AvroInt(int: Int) extends AvroValue
+  case class AvroLong(long: Long) extends AvroValue
+  case class AvroEnumSymbol(symbol: GenericEnumSymbol[_]) extends AvroValue
+  case class AvroList(list: List[AvroValue]) extends AvroValue
+  case class AvroMap(map: Map[String, AvroValue]) extends AvroValue
+  case object AvroNull extends AvroValue
+
+  case class AvroRecord(record: GenericRecord) extends AvroValue {
```

Comment:
If you declare it as a `case class`, it is not possible to make `record` private. But I think that would be desirable to hide the Java nastiness away from where it's not needed.

Reply:
It was nicer to pattern match which is why I did it. Might there be a use case to have access to the actual record?

Reply:
Well, you might have already written some code to parse a data structure from a `GenericRecord`, and if you want to reuse that, you're going to need access to the `GenericRecord`.

My original thought was that we want to keep `GenericRecord` out of the hands of users, because they might stow it away and read it later, and then they would get weird results if the `GenericRecord` was overwritten (#499). But thinking about it, they can do just the same thing with the `AvroRecord` itself, which would lead to the same result (unless `AvroRecord` creates a copy, which would be bad for performance reasons). So actually I think a `case class` is fine here.